brave-cxf3/src/main/java/com/github/kristofa/brave/cxf3/TracerContext.java

```diff
@@ -0,0 +1,67 @@
+package com.github.kristofa.brave.cxf3;
+
+import com.github.kristofa.brave.Brave;
+import com.github.kristofa.brave.ServerSpan;
+import com.github.kristofa.brave.ServerSpanThreadBinder;
+
+import java.util.concurrent.Callable;
+
+/**
+ * Used to handle {@link ServerSpan} retrieved from cxf message exchange.
+ */
+public class TracerContext {
+
+  private final ServerSpanThreadBinder serverSpanThreadBinder;
+  private final ServerSpan serverSpan;
+
+  TracerContext(Brave brave, ServerSpan serverSpan) {
+    this.serverSpanThreadBinder = brave.serverSpanThreadBinder();
+    this.serverSpan = serverSpan;
+  }
+
+  /**
+   * Wraps provided {@link Runnable and ensures same {@link ServerSpan} context as the thread from which the Runnable was executed.
+   * {@link ServerSpan} is taken from cxf {@link org.apache.cxf.message.Message} {@link org.apache.cxf.message.Exchange}
+   *
+   * @param command {@link Runnable}
+   * @return wrapped {@link Runnable}
+   */
+  public Runnable wrap(Runnable command) {
+    if (command == null) {
+      return null;
+    }
+
+    // we should clean this thread, ServerSpan will finish on another one
+    serverSpanThreadBinder.setCurrentSpan(null);
+
+    return () -> {
+      if (serverSpan != null) {
+        serverSpanThreadBinder.setCurrentSpan(serverSpan);
+      }
+      command.run();
+    };
+  }
+
+  /**
+   * Wraps provided {@link Callable} and ensures same {@link ServerSpan} context as the thread from which the Callable was executed.
+   * {@link ServerSpan} is taken from cxf {@link org.apache.cxf.message.Message} {@link org.apache.cxf.message.Exchange}
+   *
+   * @param command {@link Callable}
+   * @return wrapped {@link Callable}
+   */
+  public <V> Callable<V> wrap(Callable<V> command) {
+    if (command == null) {
+      return null;
+    }
+
+    // we should clean this thread, ServerSpan will finish on another one
+    serverSpanThreadBinder.setCurrentSpan(null);
```

Comment:
This part is loosing the current span (aka `previous`):
```
if (serverSpan != null) {
     serverSpanThreadBinder.setCurrentSpan(serverSpan);
}
```

So if the `previous` span is not null, than user either created one by himself (not good but possible) or it is a sync call, nothing to do here.  This is why I suggested something like that:
```

if (previous == null) {
    serverSpanThreadBinder.setCurrentSpan(serverSpan);
}
```
The `previous` should be always null in case of async call.

However, no matter what `previous`  one is, you are restoring it in the finally block:
```
serverSpanThreadBinder.setCurrentSpan(previous);
```

Hope it makes sense.